Tournament.cpp

```diff
@@ -0,0 +1,263 @@
+#include "Tournament.h"
+#include <iostream>
+#include <fstream>
+#include <algorithm>
+
+Tournament::Tournament(int choice, std::string& path, std::string& pathGroup)
+{
+  const int EURO = 1;
+  const int WORLD = 2;
+  if (choice == EURO)
+  {
+	try
+	{
+	  addParticipants(path, m_euroParticipants);
+	  setGroup(pathGroup, m_numberEuroGroup);
+	}
+	catch (const std::exception& ex)
+	{
+	  std::cout << "Error! "<< ex.what() << std::endl;
+	}
+	
+  }
+  if (choice == WORLD)
+  {
+	try
+	{
+	  addParticipants(path, m_worldCupParticipants);
+	  setGroup(pathGroup, m_numberWorldCupGroup);
+	}
+	catch (const std::exception& ex)
+	{
+	  std::cout << "Error!" << ex.what() << std::endl;
+	}
+  }
+}
+
+void Tournament::addParticipants(std::string& path, int numberParticipants) //The function reads participants from a file.
+{
+  std::ifstream participantsList;
+  participantsList.open(path);
+  if (!participantsList.is_open() and !participantsList.good())
+	throw std::exception("Unable to add participants. File not found");
+  else
+  {
+	std::string name;
+	int rank = 0;
+	double fifa_points = 0;
+	for (int i = 0; i < numberParticipants; ++i)
+	{
+	  participantsList >> name >> rank >> fifa_points;
+	  auto newTeam = std::make_shared<Team>(name, rank, fifa_points);
+	  m_participants.push_back(newTeam);
+	}
+  }
+  participantsList.close();
+}
+void Tournament::setGroup(std::string& path, const int tournamentGroups) //The function distributes the participants into groups.
+{
+  std::ifstream groupList;
+  groupList.open(path);
+  if(!groupList.is_open() and !groupList.good())
+	throw std::exception("Unable to set groups. File not found!");
+  else
+  {
+	for (int numberGroups = 0; numberGroups < tournamentGroups; ++numberGroups)
+	{
+	  auto newGroup = std::make_shared<Group>();
+	  newGroup->teams.resize(newGroup->numberTeams);
+	  std::generate(newGroup->teams.begin(), newGroup->teams.end(), [&]()
+	  {
+		std::string name;
+		groupList >> name;
+		return addTeam(name);
+	  });
+	  m_groups.push_back(newGroup);
+	}
+  }
+  groupList.close();
+}
+std::shared_ptr<Team> Tournament::addTeam(std::string& name)
+{
+  for (auto it : m_participants)
+  {
+	if (it->name == name)
+	  return it;
+  }
+}
+void Tournament::playGroupRound()
+{
+  int numberRounds = 3;
+  for (int i = 0; i < numberRounds; ++i)
+  {
+	for (auto it : m_groups)
+	{
+	  int rank = 0;
+	  if (i == 0)
```

Comment:
Think of a way to avoid these three ifs and instead expressing the logic without branching. I think there is a way. The code would become more readable and shorter (thus less error-prone).